feat: add JSON API for FDO Owner Server with OpenAPI specification

djach7 opened the pull request:
This pull request adds a JSON API for the FDO Owner Server that returns structured JSON documents while maintaining full backward compatibility with existing PEM-based clients.

  Changes

  - OpenAPI 3.0.3 Specification - Complete API specification with type-safe Go code generation
  - oapi-codegen Migration - Moved from npm-based openapi-generator to Go-native oapi-codegen v2.5.1
  - JSON Response System - All endpoints return structured JSON with proper content negotiation
  - Error Handling Consolidation - Centralized error patterns across API handlers
  - Content-Type Enforcement - Strict PEM upload validation per OpenAPI specification
  - Go Generate Workflow - Proper tool dependency management and code generation pipeline
  - Production Optimization - Eliminated debugging artifacts and fixed code smells
  - Shared Component Structure - Organized OpenAPI specs for future manufacturer/rendezvous servers

  API Endpoints Enhanced

  - /health - Server health status with version information (JSON)
  - /api/v1/ownerinfo - Owner redirect configuration management (JSON)
  - /api/v1/owner/vouchers - Ownership voucher operations (JSON)

  New Files

  - api/openapi/owner-server.yaml - OpenAPI 3.0.3 specification
  - api/openapi/generated.go - Generated Go types and server interfaces
  - api/openapi/config.yaml - oapi-codegen configuration
  - api/openapi/generate.go - go:generate implementation
  - api/openapi/README.md - Code generation documentation
  - api/openapi/shared/ - Shared components for future server specs
  - api/handlers/responses.go - JSON response utilities and error handling
  - api/handlersTest/json_api_test.go - JSON API validation tests
  - api/handlersTest/helpers_test.go - Test helper functions

  Modified Files

  - api/handlers/vouchers.go - JSON responses and simplified PEM handling
  - api/handlers/health.go - JSON health endpoint implementation
  - api/handlers/ownerinfo.go - JSON owner redirect API with error consolidation
  - api/handlers/rvinfo.go - Consistent error handling patterns
  - api/handlersTest/vouchers_test.go - Content-type validation tests
  - api/routes.go - Updated imports for generated types
  - scripts/fdo-utils.sh - Optimized utility functions for JSON API
  - Makefile - go generate workflow and tool management
  - go.mod / go.sum - oapi-codegen dependencies
  - .gitignore - Generated file handling

mmartinv commented on Makefile:
```diff
@@ -12,6 +12,53 @@ ARCH            := $(shell uname -m)
+	openapi-generator generate \
+		-i $(OPENAPI_SPEC) \
+		-g go \
+		-o $(GENERATED_DIR)/client \
```
Instead of using a Makefile target to generate the api I would prefer to use the golang native approach to generate code, please see https://github.com/oapi-codegen/oapi-codegen/?tab=readme-ov-file#for-go-124 as an example. We will need a Makefile target anyway but in this case it will run `go generate` and we will make the `build` target to depend on it.

This leads me to open the discussion about using the openapi generator tools (java) or the more native `oapi-codegen` shown in the link above.

After thinking a lot about this I think we should use the native tool as it's very well documented and used by other projects, but as I said this is open to discussion.

kgiusti commented on Makefile:
```diff
@@ -12,6 +12,41 @@ VERSION         := $(shell grep 'Version:' $(SPEC_FILE) | awk '{printf "%s", $$2
+.PHONY: generate-api
+generate-api: validate-openapi
+	@echo "Generating Go types and server code from OpenAPI specification..."
+	$$(command -v oapi-codegen || echo $$(go env GOPATH)/bin/oapi-codegen) -package openapi -generate types,chi-server -o $(GENERATED_FILE) $(OPENAPI_SPEC)
```
It seems that the generated file is dependent on the openapi_spec file.  If that's the case then if the openapi_spec file is modified do we want to re-generate the generated code?

If that's the case I'd add GENERATED_FILE to the list of the dependencies for the build target:
```
build: $(GENERATED_FILE) tidy fmt vet
```
Then make the phony generate-api rule into a real dependency rule for GENERATED_FILE:
```
$(GENERATED_FILE): $(OPENAPI_SPEC) validate-openapi
   blab...
```
You can keep the phony generate-api rule, but it would just be a dependency on $(GENERATED_FILE) with no body
```
.PHONY: generate-api
generate-api: $(GENERATED_FILE)
```

Again this is only useful if you want "make build" to re-generate the code if the spec file has changed.

kgiusti commented on api/handlers/responses.go:
```diff
@@ -0,0 +1,145 @@
+// HandleDBError handles GORM errors consistently (duplicate key or not found)
+func HandleDBError(w http.ResponseWriter, r *http.Request, entityName string, err error) bool {
+	if errors.Is(err, gorm.ErrDuplicatedKey) {
+		slog.Error(entityName+" already exists (constraint)", "error", err)
```
WriteErrorResponse?

kgiusti commented on api/handlers/vouchers.go:
```diff
@@ -58,37 +53,48 @@ func GetVoucherHandler(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
 		if errors.Is(err, gorm.ErrRecordNotFound) {
-			http.Error(w, "Voucher not found", http.StatusNotFound)
+			if ShouldReturnJSON(r) {
```
WriteErrorResponse?

kgiusti reviewed (commented):
Couple 'o questions:  First, why doesn't the patch use the generated ServerInterface/Handler code?  Seems like we're not leveraging a big chunk of the generated functionality.  When I look at the server examples on the oapi-codegen [website](https://github.com/oapi-codegen/oapi-codegen?tab=readme-ov-file#go-122-nethttp) it appears that we only need to provide the ServiceInterface impl and the Http server setup code.

Second, what's the reason the patch uses the chi server as opposed to the go-native (or other server options)?  Not implying it's wrong or anything - I'm not familiar with these other server impls - just interested in the motivation for choosing chi.  Thanks

mmartinv commented on .github/workflows/ci.yml:
```diff
@@ -23,6 +23,9 @@ jobs:
         uses: actions/checkout@v4
 
+      - name: Install oapi-codegen
+        run: go install github.com/oapi-codegen/oapi-codegen/v2/cmd/oapi-codegen@latest
```
I would move this to a `Makefile` target.

mmartinv commented on .gitignore:
```diff
@@ -6,3 +6,6 @@ go-fdo-server
 test/workdir
+
+# Generated OpenAPI code
+generated/
```
We need to track the changes in the generated code, we just cannot ignore the generated files

mmartinv commented on Makefile:
```diff
@@ -12,9 +12,47 @@ VERSION         := $(shell grep 'Version:' $(SPEC_FILE) | awk '{printf "%s", $$2
+.PHONY: clean-generated
+clean-generated:
+	@echo "Cleaning generated OpenAPI code..."
+	rm -f $(GENERATED_FILE)
```
This target is not needed I believe

djach7 commented on Makefile:
```diff
@@ -12,9 +12,47 @@ VERSION         := $(shell grep 'Version:' $(SPEC_FILE) | awk '{printf "%s", $$2
 # Build the Go project
 .PHONY: build
-build: tidy fmt vet
+build: $(GENERATED_FILE) tidy fmt vet
```
I implemented this in the most recent change but please take another look to make sure I've got the right idea.

mmartinv commented on test/ci/utils.sh:
```diff
@@ -480,12 +481,12 @@ set_or_update_owner_redirect_info() {
   else
     log_info "'RVTO2Addr' found, updating it"
-    update_owner_redirect_info "${owner_url}" "${real_owner_ip}" "${owner_dns}" "${owner_port}" "${owner_protocol}"
+    update_ownerinfo "${owner_url}" "${real_owner_ip}" "${owner_dns}" "${owner_port}" "${owner_protocol}"
```
It's OK if you rename the functions but do the same in `scripts/fdo-utils.sh`: you added new functions not renamed and modified the existing ones.

mmartinv commented on .github/workflows/ci.yml:
```diff
@@ -23,6 +23,7 @@ jobs:
       - name: Check out repository code
         uses: actions/checkout@v4
 
+
```
Remove all the unwanted new lines in all the files

djach7 commented:
Addressed in the most recent commit, I removed the paths and components from the owner spec and created a new `shared-types.go` generated file to grab those from.

kgiusti commented on api/openapi/owner-server.yaml:
```diff
@@ -0,0 +1,104 @@
+      type: array
+      items:
+        type: object
+        required: [dns, port, protocol]
```
There's an "ip" entry missing here.  See https://github.com/fido-device-onboard/go-fdo-server/blob/d37e23dca1ea2d601d35911973a8c7f0f3ade513/internal/db/db.go#L511

kgiusti commented on api/openapi/generate.go:
```diff
@@ -0,0 +1,6 @@
+package openapi
+
+//go:generate go tool oapi-codegen -config config.yaml owner-server.yaml
```
I would recommend that you try to generate the owner server code under its own owner package. In other words:
```
oapi-codegen -config config.yaml -package oapi_owner owner-server.yaml
```

We will need this eventually anyways in order to support three different ServiceInterface interfaces, as the REST API for each server will be different. 

With that each server is allowed to define a custom Server type tailored to its needs, which it will then pass to the generated Handler() function imported from its server-specific package.

The common types should continue to use the "openapi" package name since... well, they're common.

djach7 replied:
I refactored to generate the owner server code in its own package, should be separated now

djach7 commented on cmd/owner.go:
```diff
@@ -320,15 +320,29 @@ func serveOwner(config *OwnerServerConfig) error {
+	mainHandler.Handle("POST /fdo/101/msg/{msg}", handler)
+
+	// Register ownerinfo endpoint manually (not part of OpenAPI spec but needed for TO0)
+	mainHandler.HandleFunc("/api/v1/ownerinfo", handlers.OwnerInfoHandler)
```
My understanding was that the endpoints server different purposes. The `ownerinfo` endpoint is required for FDO TO0, and stores a JSON blob in the `owner_info` table. The `redirect` endpoint manages owner redirect configurations and provides the new JSON API interface for consumers.

djach7 commented on api/handlers/ownerinfo.go:
```diff
@@ -23,86 +20,66 @@ func OwnerInfoHandler(w http.ResponseWriter, r *http.Request) {
 
-func getOwnerInfo(w http.ResponseWriter, _ *http.Request) {
-	slog.Debug("Fetching ownerInfo")
+func getOwnerInfo(w http.ResponseWriter, r *http.Request) {
```
I think you are 100% correct, the ownerinfo functions were unnecessary and removed in the recent changes. As for the second bit, @runcom would you like me to work towards folding your PR (#118) into this one? I don't mind either way, I can also create a follow up PR later to make naming adjustments too.

mmartinv commented on api/components.yaml:
```diff
@@ -0,0 +1,67 @@
+            items:
+              oneOf:
+                - type: integer
+                - type: string
```
This definition is wrong and seems to correspond to the rvinfo not to the rvto2addr, please see https://github.com/fido-device-onboard/go-fdo-server/pull/154/changes#diff-2e7681f1cfb4420391185bff11a9619926890b4b81b8558b59ac964c251e5628R92-R113

mmartinv commented on api/rvto2addr.yaml:
```diff
@@ -0,0 +1,59 @@
+  description: Rendezvous TO2 address management endpoints
+
+paths:
+  /owner/redirect:
```
```suggestion
  /rvto2addr:
```

mmartinv commented on internal/db/state.go:
```diff
@@ -98,6 +99,68 @@ func (s *State) Close() error {
 	return sqlDB.Close()
 }
 
+// FetchRvInfoJSON reads the rvinfo JSON from the database
```
I think you are mixing things here: RVInfo is not part of the owner API.
